File: solutions/C++/remove_duplicates_from_sorted_array_simple.cpp
#include <iostream>
#include <vector>

int removeDuplicates(std::vector<int>& nums) {
    /*
    Simple Pass Approach
    Remove duplicates by iterating and comparing consecutive elements.

    Time Complexity: O(n)
    Space Complexity: O(1)
    */
    if (nums.empty()) {
        return 0;
    }

    int write_idx = 0;
    for (int read_idx = 1; read_idx < (int)nums.size(); read_idx++) {
        if (nums[read_idx] != nums[write_idx]) {
            write_idx++;
            nums[write_idx] = nums[read_idx];
        }
    }

    return write_idx + 1;
}

int main() {
    std::vector<int> nums1 = {1, 1, 2};
    std::cout << removeDuplicates(nums1) << std::endl;  // 2, nums = [1, 2, _]

    std::vector<int> nums2 = {0, 0, 1, 1, 1, 2, 2, 3, 3, 4};
    std::cout << removeDuplicates(nums2) << std::endl;  // 5, nums = [0, 1, 2, 3, 4, _, _, _, _, _]

    return 0;
}

File: solutions/Golang/remove_duplicates_from_sorted_array_simple.go
package main

import "fmt"

/*
Simple Pass Approach
Remove duplicates by iterating and comparing consecutive elements.

Time Complexity: O(n)
Space Complexity: O(1)
*/
func removeDuplicates(nums []int) int {
    if len(nums) == 0 {
        return 0
    }

    writeIdx := 0
    for readIdx := 1; readIdx < len(nums); readIdx++ {
        if nums[readIdx] != nums[writeIdx] {
            writeIdx++
            nums[writeIdx] = nums[readIdx]
        }
    }

    return writeIdx + 1
}

func main() {
    nums1 := []int{1, 1, 2}
    fmt.Println(removeDuplicates(nums1))  // 2, nums = [1, 2, _]

    nums2 := []int{0, 0, 1, 1, 1, 2, 2, 3, 3, 4}
    fmt.Println(removeDuplicates(nums2))  // 5, nums = [0, 1, 2, 3, 4, _, _, _, _, _]
}
